File: composeApp/src/androidMain/kotlin/io/github/a7asoft/ui/components/StatusHeader.kt
package io.github.a7asoft.ui.components

import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.material3.Button
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.text.style.TextAlign
import androidx.compose.ui.unit.dp
import io.github.a7asoft.ui.ReaderUiState

/**
 * Top region of the reader screen. Shows a state-driven headline and
 * subtitle, plus an "Open NFC settings" CTA when NFC is disabled.
 */
@Composable
public fun StatusHeader(
    state: ReaderUiState,
    onOpenNfcSettings: () -> Unit,
) {
    Column(
        modifier = Modifier.fillMaxWidth(),
        horizontalAlignment = Alignment.CenterHorizontally,
        verticalArrangement = Arrangement.spacedBy(12.dp),
    ) {
        Text(
            text = headlineFor(state),
            style = MaterialTheme.typography.headlineSmall,
            textAlign = TextAlign.Center,
        )
        Text(
            text = subtitleFor(state),
            style = MaterialTheme.typography.bodyMedium,
            textAlign = TextAlign.Center,
        )
        if (state is ReaderUiState.NfcDisabled) {
            Button(onClick = onOpenNfcSettings) { Text("Open NFC settings") }
        }
    }
}

@Suppress("CyclomaticComplexMethod")
// why: exhaustive when over sealed UI state; intermediate progress states
// collapse via `else ->`. Each branch is a single mapping.
private fun headlineFor(state: ReaderUiState): String = when (state) {
    ReaderUiState.NfcUnavailable -> "NFC not supported"
    ReaderUiState.NfcDisabled -> "NFC is off"
    ReaderUiState.Idle -> "Tap a card"
    is ReaderUiState.Done -> "Card read"
    is ReaderUiState.Failed -> "Read failed"
    else -> "Reading"
}

@Suppress("CyclomaticComplexMethod")
// why: exhaustive when over sealed UI state; intermediate progress states
// collapse via `else ->`. Each branch is a single mapping.
private fun subtitleFor(state: ReaderUiState): String = when (state) {
    ReaderUiState.NfcUnavailable -> "This device has no NFC hardware. The reader cannot run here."
    ReaderUiState.NfcDisabled -> "Enable NFC in system settings to read a card."
    ReaderUiState.Idle -> "Hold a contactless EMV card against the back of the device."
    is ReaderUiState.Done -> "Hold another card to read again."
    is ReaderUiState.Failed -> "Try again, or use a different card."
    else -> "Hold the card steady"
}

File: iosApp/iosApp/ui/components/StatusHeader.swift
import SwiftUI

/// Top region of the reader screen. Shows a state-driven headline and
/// subtitle so the user always knows what the app is doing.
///
/// Mirrors the Android composeApp `StatusHeader` discipline: the
/// `headline` / `subtitle` switches are exhaustive over
/// ``ReaderUiState`` so adding a new case forces a compile error here.
public struct StatusHeader: View {

    private let state: ReaderUiState

    /// Build a header bound to the current ``ReaderUiState``.
    public init(state: ReaderUiState) {
        self.state = state
    }

    public var body: some View {
        VStack(spacing: 12) {
            Text(headline)
                .font(.title2)
                .multilineTextAlignment(.center)
            Text(subtitle)
                .font(.body)
                .foregroundColor(.secondary)
                .multilineTextAlignment(.center)
        }
        .frame(maxWidth: .infinity)
    }

    private var headline: String {
        switch state {
        case .nfcUnavailable: return "NFC not supported"
        case .idle: return "Tap a card"
        case .scanning: return "Hold a card"
        case .done: return "Card read"
        case .failed: return "Read failed"
        case .tagDetected, .selectingPpse, .selectingApplication, .readingRecords: return "Reading"
        }
    }

    private var subtitle: String {
        switch state {
        case .nfcUnavailable: return "This device has no NFC reader. The sample cannot run here."
        case .idle: return "Press Scan and hold a contactless EMV card against the top of the device."
        case .scanning: return "Hold the card flat against the top of the device."
        case .done: return "Press Read another card to scan again."
        case .failed: return "Press Try again, or use a different card."
        case .tagDetected, .selectingPpse, .selectingApplication, .readingRecords: return "Hold the card steady"
        }
    }
}
